jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/internal/UpgradeResponseDelegate.java

```diff
@@ -64,14 +63,7 @@ public Set<String> getHeaderNames()
     @Override
     public Map<String, List<String>> getHeaders()
     {
-        Map<String, List<String>> result = new LinkedHashMap<>();
-        HttpFields.Mutable headers = response.getHeaders();
-        for (HttpField header : headers)
-        {
-            String name = header.getName();
-            result.put(name, headers.getValuesList(name));
-        }
-        return result;
+        return Collections.unmodifiableMap(HttpFields.asMap(response.getHeaders()));
```

Comment:
Because this is not used in the negotiation but stored in the `FrameHandler` after websocket upgrade. So it makes no sense for them to alter the HttpHeaders after the websocket upgrade is completed.

Reply:
I think it is assuming too much of what the callers of this method are doing.
If the caller needs it unmodifiable, should be the caller wrapping it.

Besides, should not all the other methods wrap as unmodifiable?
For example, `getHeaderNames()`, getHeaders(String)`?

I'm fine to make everything unmodifiable in this class, since it's only used in 1 place, or have the 1 caller to make everything unmodifiable, but not half/half.

Reply:
> Besides, should not all the other methods wrap as unmodifiable?
> For example, getHeaderNames(), getHeaders(String)`?

The difference is that `getHeaders().put(....)` will add actually modify the `HttpFields` on the response.
But `getHeaderNames()` and `getHeaders(String)` will return a new `Set` or `List` and so will not change the underlying `HttpFields` if you add to them.

Anyway, I removed the unmodifiableMap... But it is completely wrong for the application to modify the headers of the `Response` after we have fully upgraded to WebSocket.

Reply:
The response header fields are switched internally to be immutable when the response is committed.  So this should be safe.